nikola/filters.py

```diff
@@ -437,3 +437,38 @@ def add_header_permalinks(data, xpath_list=None):
             new_node = lxml.html.fragment_fromstring('<a href="#{0}" class="headerlink" title="Permalink to this heading">¶</a>'.format(hid))
             node.append(new_node)
     return lxml.html.tostring(doc, encoding="unicode")
+
+
+@apply_to_text_file
+def deduplicate_ids(data):
+    """Post-process HTML via lxml to deduplicate IDs."""
+    doc = lxml.html.document_fromstring(data)
+    elements = doc.xpath('//*')
+    all_ids = [element.attrib.get('id') for element in elements]
+    seen_ids = set()
+    duplicated_ids = set()
+    for i in all_ids:
+        if i is not None and i in seen_ids:
+            duplicated_ids.add(i)
+        else:
+            seen_ids.add(i)
+
+    if duplicated_ids:
+        # Well, that sucks.
+        for i in duplicated_ids:
+            # Results are ordered the same way they are ordered in document
+            offending_elements = doc.xpath('//*[@id="{}"]'.format(i))
+            counter = 2
+            for e in offending_elements[1::-1]:
+                new_id = '{0}-{1}'.format(i, counter)
```

Comment:
But trying until you find a free one will be a problem if you want to make permalinks (as for the Sphinx permalinks filter). Assume the two oldest entries have headers with IDs `a` and `a`. If there's nothing else around, one will get `a` and the other `a-2`. But now assume a new post is added which uses `a-2` (for whatever reason). Then the old `a-2` will end up as `a-3`, because `a-2` is already taken, and links pointing to `a` in the second oldest post are suddenly pointing to the wrong place.